pkg/s3wrapper/handler.go

```diff
@@ -0,0 +1,56 @@
+package s3wrapper
+
+import (
+	"github.com/openshift/assisted-service/internal/isoeditor"
+	"github.com/openshift/assisted-service/internal/versions"
+	"github.com/sirupsen/logrus"
+)
+
+// NewStorageClient Create new storage client based off the inputs of DEPLOY_TARGET and STORAGE env variables
+func NewStorageClient(deployTarget string, storage string, s3cfg *Config, fsWorkDir string, log logrus.FieldLogger, versionsHandler versions.Handler, isoEditorFactory isoeditor.Factory) API {
+	var storageClient API
+	if storage != "" {
+		switch storage {
+		case "s3":
+			storageClient = NewS3Client(s3cfg, log, versionsHandler, isoEditorFactory)
+			if storageClient == nil {
+				log.Fatal("failed to create S3 client")
+			}
+		case "filesystem":
+			storageClient = NewFSClient(fsWorkDir, log, versionsHandler, isoEditorFactory)
+			if storageClient == nil {
+				log.Fatal("failed to create filesystem client")
+			}
+		default:
+			log.Fatalf("unsupported storage client: %s", storage)
+		}
+	} else {
+		switch deployTarget {
+		case "k8s":
+			storageClient = NewS3Client(s3cfg, log, versionsHandler, isoEditorFactory)
+			if storageClient == nil {
+				log.Fatal("failed to create S3 client")
+			}
+		case "onprem", "ocp":
+			storageClient = NewFSClient(fsWorkDir, log, versionsHandler, isoEditorFactory)
+			if storageClient == nil {
+				log.Fatal("failed to create S3 filesystem client")
+			}
+		default:
+			log.Fatalf("unsupported deploy target %s", deployTarget)
+		}
+	}
+	return storageClient
+}
+
+// CreateS3BucketFromEnv Create bucket if the env variable CREATE_S3_BUCKET is true
+func CreateS3BucketFromEnv(envflag bool, log logrus.FieldLogger, objectHandler API) {
+	if envflag {
```

Comment:
Can this conditional be in the caller? Seems strange to have a no-op based on a parameter ... just don't call the function and get rid of the parameter.

Reply:
I'm not sure. I was really just reacting to the way this `envflag` parameter is being handled.

Now that I'm thinking about this particular function though, I think it belongs outside of `s3wrapper`.
I would prefer to keep `s3wrapper` as close to just a client library as we can, which means keeping application-specific knowledge ("we need a regular and a public bucket") out of this package.

Reply:
Also I think it's especially unexpected to `log.Fatal` in a package function as that exits the program. We should only ever exit from main.

Reply:
It's the same helper function as in main I just renamed Options.CreateS3Bucket to envflag as it is passed in. It seems to be just to avoid a copy pasta for this logic in the Options.deployTarget switch cases. I can put it back there :)

I don't wish to complicate things much, but it just seemed related to me because it is a storage option being configured. My gut is that it's strange to me that the s3wrapper.createS3Bucket() functions don't take into account this env variable and instead are wrapped by this function when they are called. Maybe makes sense to put that logic for the env variable in the createS3Bucket functions?

Reply:
Reverted this change